src/full/Agda/Interaction/JSONTop.hs

```diff
@@ -0,0 +1,203 @@
+-- {-# LANGUAGE CPP #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+
+module Agda.Interaction.JSONTop
+    ( jsonREPL
+    ) where
+import Control.Monad.State
+
+import Data.Aeson hiding (Result(..))
+import Data.ByteString.Lazy (ByteString)
+import qualified Data.ByteString.Lazy.Char8 as BS
+import Data.Char
+import qualified Data.List as List
+import Data.Maybe
+
+import System.IO
+
+import Agda.Utils.Monad
+import Agda.Utils.Maybe
+import Agda.Utils.Pretty
+import Agda.Utils.String
+
+import Agda.Syntax.Common
+
+import Agda.TypeChecking.Monad
+import qualified Agda.TypeChecking.Monad.Benchmark as Bench
+
+import Agda.Interaction.Response as R
+import Agda.Interaction.InteractionTop
+import Agda.Interaction.EmacsCommand hiding (putResponse)
+import Agda.Interaction.Highlighting.JSON
+import Agda.Interaction.Highlighting.Precise (TokenBased(..))
+import Agda.Interaction.Options
+
+import Agda.VersionCommit
+
+----------------------------------
+
+-- | 'jsonREPL' is a interpreter like 'mimicGHCi', but outputs JSON-encoded strings.
+--
+--   'jsonREPL' reads Haskell values (that starts from 'IOTCM' ...) from stdin,
+--   interprets them, and outputs JSON-encoded strings. into stdout.
+
+jsonREPL :: TCM () -> TCM ()
+jsonREPL setup = do
+    liftIO $ do
```

Comment:
Same here. Looks like we could define an `agdaTop` which would
subsume `jsonREPL` and `mimicGHCi`. It seems to only need to take
as an argument the callback function and the prompt string.